### DISCUSSION TITLE
Using getInitialProps in _app.js + getStaticProps in pages/*.js breaks data returned by getInitialProps

### DISCUSSION THREAD

**Discussion by jeffweim-hm** [vercel/next.js#30967] [Help]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

14.16.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

I'm using `getInitialProps` within my `_app.js` file to get global user data needed for all pages. Within my individual pages, I use `getStaticProps` to then get each page's set of data.

When I combine these 2 Next.js APIs together, the data I am attempting to return from `getInitialProps` never makes it to `MyApp`. It's always the default props defined using `PropTypes.` However, if I remove `getStaticProps` from any one of the pages, I can see the data returned from `getInitialProps` within `MyApp`'s props.

This seems to be a bug as I do not see anywhere in the documentation that these 2 APIs cannot be used together.

I should also note, that this only happens when running `next start`, but does not happen when running `next dev`.

If this is not a bug, how can I use these 2 APIs together and still get the data returned from `getInitialProps` to be available to `MyAPP`?

`_app.js`

```javascript
const MyApp = (props) => {
  const {
    Component,
    pageProps,
    user,

  } = props

  return (
    <AuthProvider
      value={{
        user,
      }}
    >
      <Component {...pageProps} />
    </AuthProvider>
  )
}

MyApp.getInitialProps = async (appContext) => {
  const myCookie = appContext?.ctx?.req?.cookies?.myCookie

  if (myCookie) {
    try {
      const {
        data: { user },
      } = await axios.post(`${ process.env.URL_VALUE }/api/auth`, { myCookie })

      return {
        user,
      }
    } catch (error) {
      console.log(error)

      return {}
    }
  }
  return {}
}
```

### Expected Behavior

The data being returned from `getInitialProps` should be available to `MyApp`. In this case, the data is not there. 

### To Reproduce

1. Create custom `_app.js` using `getInitialProps`
2. Return some kind of data from `getInitialProps`
3. Create a page, `index.js` works
4. In `index.js` use `getStaticProps` to return some kind of page-specific data
5. View whether the data returned from `getInitialProps` is available

**Comment by ijjk** [comment]
Hi, it looks like the original `getInitialProps` isn't being called as shown in the documentaiton [here](https://nextjs.org/docs/advanced-features/custom-app).

**Reply by jeffweim-hm** [reply]
@ijjk I've attempted the use of `getInitialProps` using the setup from the docs, with the same (not working) results. I had tried using this before with no luck, which is why I didn't mention having used it in my original description. 

My `_app.js` usage with the information from the docs:
```javascript
const MyApp = (props) => {
  const {
    Component,
    pageProps,
    user,

  } = props

  return (
    <AuthProvider
      value={{
        user,
      }}
    >
      <Component {...pageProps} />
    </AuthProvider>
  )
}

MyApp.getInitialProps = async (appContext) => {
  const myCookie = appContext?.ctx?.req?.cookies?.myCookie
  const appProps = await App.getInitialProps(appContext)

  if (myCookie) {
    try {
      const {
        data: { user },
      } = await axios.post(`${ process.env.URL_VALUE }/api/auth`, { myCookie })

      return {
        ...appProps,
        user,
      }
    } catch (error) {
      console.log(error)

      return {
        ...appProps
      }
    }
  }
  return {
    ...appProps
  }
}
```

**Reply by ijjk** [reply]
I'm not able to reproduce this on the latest version of Next.js `v12.0.2`, are you using `revalidate` on your `getStaticProps` pages? If not `getStaticProps` will only be called during build-time and won't be called during runtime as the build cache is used instead.

**Reply by jeffweim-hm** [reply]
@ijjk I was not using `revalidate` in the return of `getStaticProps`. It does seem to work partially. Initial load of the page, user data is not there, on a secondary reload, it is.

I'm setting it to `revalidate: true`. Is this the expected behavior of this option?

**Reply by callmegin** [reply]
@ijjk I'm actually facing similar issue like @jeffweim-hm  as well, while using Next-Auth. I've posted a question in[ Next-Auth discussion](https://github.com/nextauthjs/next-auth/discussions/3890) with more details and link to [example project repo](https://github.com/callmegin/next-auth-test), though nobody replied yet. Project should be compiled in order to reproduce.

In addition to data being available after a secondary reload, it is available in private tab as well after first reload, though second time - it's not, how it should be in the first place I imagine ?

![next auth getstaticprops](https://user-images.githubusercontent.com/70571729/153225392-46b98585-6bea-4829-8cb1-8bfc05871768.gif)

Any info would be greatly appreciated.

**Comment by 2stash** [comment]
@jeffweim-hm I am building out an app and I am adding getInitialProps in my _app.js file similar to how you have it.  I was able to get getInitialProps and getStaticProps to work on my local dev server. The thing I noticed was naming convention clashes and passing down the props explicitly.

```
function MyApp({Component, pageProps, info}){
return (
<Layout info={info}> 
<Component {...pageProps} {...info}/>
</Layout>
)}

MyApp.getInitialProps = async(ctx) => {
const info = await fetch("api/api/api")
return {
info: {info}
}}


```

Then my getStaticProps in my page file returns props.

```
//getStaticProps function

const data = fetch("api")
return {
props:{data}
}

```
Not sure if this helps or I just haven't run into the problem yet. My layout is a sidebar and nav that I need to populate with some data from the api call, which is why I am trying to use getInitialProps in the _app.js component.  I am able to console.log the data from the 2 api calls in the components themselves so the api calls are working and the data is passed to the components. I was checking the forums for any useful info on getInitialProps and saw this post, so if I run into an issue later I'll comment here.
